build_tools/provision.py

```diff
@@ -0,0 +1,231 @@
+            if amdgpu_family in asset["name"]:
+                return asset
+
+    return None
+
+
+def _download_github_release_asset(asset_data, output_dir):
+    """
+    With the GitHub asset data, this function downloads the asset to the output_dir
+    """
+    asset_name = asset_data["name"]
+    asset_url = asset_data["url"]
+    destination = Path(output_dir) / asset_name
+    headers = {"Accept": "application/octet-stream"}
+    # Making the API call to retrieve the asset
+    response = requests.get(asset_url, stream=True, headers=headers)
+
+    # Downloading the asset in chunks to destination
+    # In order to get better visibility on downloading status, tqdm adds a progress bar
+    total_size = int(response.headers.get("content-length", 0))
+    block_size = 1024
+    with tqdm(total=total_size, unit="B", unit_scale=True) as progress_bar:
+        with open(destination, "wb") as file:
+            for chunk in response.iter_content(block_size * block_size):
+                progress_bar.update(len(chunk))
+                file.write(chunk)
+
+    # After downloading the asset, untar-ing the file
+    _untar_files(output_dir, destination)
+
+
+def retrieve_artifacts_by_run_id(args):
+    """
+    If the user requested TheRock artifacts by CI (run ID), this function will retrieve those assets
+    """
+    run_id = args.run_id
+    output_dir = args.output_dir
+    amdgpu_family = args.amdgpu_family
+    log(f"Retrieving artifacts for run ID {run_id}")
+    if not s3_bucket_exists(run_id):
+        log(f"S3 artifacts for {run_id} does not exist. Exiting...")
+        return
+
+    args.all = True
+
+    # Retrieving base and all math-lib tar artifacts and downloading them to output_dir
+    retrieve_base_artifacts(args, run_id, output_dir)
+    retrieve_enabled_artifacts(args, True, amdgpu_family, run_id, output_dir)
+
+    # Flattening artifacts from .tar* files then removing .tar* files
+    log(f"Untar-ing artifacts for {run_id}")
+    output_dir_path = Path(output_dir).resolve()
+    tar_file_paths = list(output_dir_path.glob("*.tar.*"))
+    flattener = ArtifactPopulator(
+        output_path=output_dir_path, verbose=True, flatten=True
+    )
+    flattener(*tar_file_paths)
+    for file_path in tar_file_paths:
+        file_path.unlink()
+
+    log(f"Retrieved artifacts for run ID {run_id}")
+
+
+def retrieve_artifacts_by_release(args):
+    """
+    If the user requested TheRock artifacts by release tag (github.run_id), this function will retrieve those assets
+    """
+    output_dir = args.output_dir
+    amdgpu_family = args.amdgpu_family
+    # Searching for X.X.XrcYYYYMMDD format to see if this is a nightly-release or dev-release
+    regex_expression = "(\\d+\\.)?(\\d+\\.)?(\\*|\\d+)rc(\\d{4})(\\d{2})(\\d{2})"
+    nightly_release = re.search(regex_expression, args.release) != None
+    release_tag = "nightly-release" if nightly_release else "dev-release"
+    release_version = args.release
```

Comment:
Please don't create a regex yourself. [packaging.version](https://packaging.pypa.io/en/stable/version.html) has all you need, in particular `is_prerelease` and `is_devrelease`.